### PR TITLE
Issue #14805: Remove support for string templates syntax

### LINKED ISSUE #14805: Remove Support for String Template Syntax
We always take a risk supporting preview features, and try to wait until second/third/final preview to add support for new syntax, to improve the chance that a given feature will become a part of the language permanently. Of course, this is also driven by user demand.

Unfortunately, String Templates have been completely removed from JDK 23 (they are not even a preview feature). Since we have really only begun to implement them (text block templates are not supported, and no checks have really been updated), it is better to remove support completely now, than to end up supporting syntax that is not valid and updating a bunch of checks to only have to rewrite the grammar, build the AST, and update the checks all over again once the syntax is finalized.

**Enabling us to start with a clear slate when the new syntax is revealed will help us to deliver support much faster.**

It is doubtful that whatever the new (post JDK 23) String Template syntax will look the same, so we will have to break compatibility with the JDK 22 String Template syntax regardless.

From  https://mail.openjdk.org/pipermail/amber-spec-experts/2024-April/004106.html:

>So, to be clear: there will be no string template feature, even with --enable-preview, in JDK 23.

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #14805

-----------------------

### Reports
Projects file: projects-to-test-on.properties
Patch branch: pull-15212

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/part1/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/part4/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/part5/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/part3/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/part2/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/part6/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/sevntu-check-regression_part_1/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/sevntu-check-regression_part_2/index.html

All previous reports are good. differences are:
-  removal of violations in files that have string template
-  line numbers in exception messages ([VarargsAndReceiver](https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/sevntu-check-regression_part_2/openjdk21/index.html#A195) file)

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/antlr-report-checkstyle/index.html

Report is good difference are:
- String templates trees are removed and printing exception message instead
- line numbers in exception messages

I just don't understand [this](https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/pull-15212/2024-07-10-T-00-20-42/antlr-report-checkstyle/checkstyle/index.html#A2220). Is this expected? it is still stackoverflow but the difference in messages is not just line numbers

### CODE REVIEW

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 [hunk 1]
```diff
@@ -116,12 +116,12 @@ tokens {
     STRING_TEMPLATE_CONTENT, EMBEDDED_EXPRESSION_BEGIN, EMBEDDED_EXPRESSION,
     EMBEDDED_EXPRESSION_END,
```
Comment (mahfouz72) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:118]: had to leave string templates token in the lexer due to: 
```
[ERROR]   GeneratedJavaTokenTypesTest.testTokenNumbering:741 A token's number has changed. Please open 'GeneratedJavaTokenTypesTest' and confirm which token is at fault.
Token numbers must not change or else they will create a conflict with users.
See Issue: https://github.com/checkstyle/checkstyle/issues/505
```
If we remove string template tokens from the lexer the last two tokens' numbers will change and we have a test that ensures that the tokens' numbers don't change. 

accordingly, coverage for javaLnaguageLexer.java will drop 

@nrmancuso please help in this how to remove the tokens and pass GeneratedJavaTokenTypesTest.testTokenNumbering ?

**Comment by romani** [comment]
What if we put `__` to make it clear it is not for use at all ? Or something like this.
Deprecated might be considered as part of language terminology.

**Comment by mahfouz72** [comment]
I had to use `DEPRECATED_` as a prefix instead of `__`
because it is a syntax error if name starts with underscore

```
[ERROR] error(50): JavaLanguageLexer.g4:119:4: syntax error: '_' came as a complete surprise to me 
```

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 [hunk 2]
```diff
@@ -112,16 +112,19 @@ tokens {
...
     RECORD_PATTERN_DEF, RECORD_PATTERN_COMPONENTS,
 
-    STRING_TEMPLATE_BEGIN, STRING_TEMPLATE_MID, STRING_TEMPLATE_END,
-    STRING_TEMPLATE_CONTENT, EMBEDDED_EXPRESSION_BEGIN, EMBEDDED_EXPRESSION,
-    EMBEDDED_EXPRESSION_END,
+    // We removed support for string template. String template tokens
+    // are no longer used, but we need to keep them to preserve token numbering,
+    // so that we are not breaking change with older versions.
+    // We used 'DEPRECATED' as prefix to indicate that they are not used.
+    DEPRECATED_STRING_TEMPLATE_BEGIN,DEPRECATED_STRING_TEMPLATE_MID, DEPRECATED_STRING_TEMPLATE_END,
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:119]: __Not__For_usage1__, __Not__For_usage2__
To completely forget why it was added.

I open for better names.

Comment (mahfouz72) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:119]: good 
@nrmancuso what do you think

Comment (nrmancuso) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:119]: We can do `NOT_FOR_USAGE_1` with comment that these are placeholders to maintain token ordering, we can skip mentioning string templates

Comment (mahfouz72) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4]: done

**Comment by rnveach** [comment]
> we consider renumbering of tokens in our API as a breaking change, so we need to keep these tokens unless we can agree that this case is exceptional

Isn't this whole PR a breaking change? We are removing what we once supported. Issue is already labeled as this.

I am good to remove all existence of the old tokens and change numbers. I prefer a clean break over leaving crazyness in our code.

I think this should be why we shouldn't support any preview features until they get out of the preview state.

**Comment by nrmancuso** [comment]
> I am good to remove all existence of the old tokens and change numbers. I prefer a clean break over leaving crazyness in our code.

I am good, @romani?

**Comment by romani** [comment]
Yes, it is breaking changes.
But there is more problems, such int values are inclined during compilation to others third party and not changing. If some plugin/extension in compiled with 10.17.0 and executed with some 11.x where we will have same numbers, it will be disaster to figure out a problem.

We already did this ones in a past :). Users were upset. Let's not deal with them. Managing API is hard , and we need to do it in hard way. Fortunately this deprecation is easy.
Such numbers will never be used, custom Checks will not be called :), no Check execution, no violations, easy to detect anomalies.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 [hunk 3]
```diff
@@ -112,16 +112,16 @@ tokens {
     PERMITS_CLAUSE, PATTERN_DEF, LITERAL_WHEN,
     RECORD_PATTERN_DEF, RECORD_PATTERN_COMPONENTS,
 
-    STRING_TEMPLATE_BEGIN, STRING_TEMPLATE_MID, STRING_TEMPLATE_END,
-    STRING_TEMPLATE_CONTENT, EMBEDDED_EXPRESSION_BEGIN, EMBEDDED_EXPRESSION,
-    EMBEDDED_EXPRESSION_END,
+    // Placeholders for tokens that are no longer used, But they
+    // have been kept here to maintain compatibility in token numbering
+    NOT_FOR_USAGE_1,NOT_FOR_USAGE_2, NOT_FOR_USAGE_3, NOT_FOR_USAGE_4,
```
Comment (nrmancuso) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:117]:
```suggestion
    NOT_FOR_USAGE_1, NOT_FOR_USAGE_2, NOT_FOR_USAGE_3, NOT_FOR_USAGE_4,
```

**Review by nrmancuso — CHANGES_REQUESTED** [review]
minor item:

**Comment by nrmancuso** [comment]
I am prepared to approve after is completed, and reports at https://github.com/nrmancuso/checkstyle-diff-report-generator/actions/runs/9864962646 look good.

**Comment by nrmancuso** [comment]
To be clear, nothing blocks us from making a new issue to remove the placeholder tokens and breaking compatibility, @rnveach feel free to open such an issue and we can continue the discussion there. We need to get this PR across the finish line.

**Comment by nrmancuso** [comment]
@romani please proceed with review

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 (hunk 3 above)
Comment (mahfouz72) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4]: done

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by nrmancuso** [comment]
@mahfouz72 I have added the reports to the PR description, please analyze them and share results for each under each report.

**Review by nrmancuso — APPROVED** [review]
Approved conditionally based on report results.

**Comment by nrmancuso** [comment]
Reports look good, the stack overflow causes undefined behavior which can manifest on our side or ANTLR’s side.

**Comment by mahfouz72** [comment]
```
Execution Time Baseline: 415.72 s
Average Execution Time: 384.02 s
============================================================
Execution Time Difference: -7.6200%
```

interesting statistics, but why is there  a huge difference compared to other PR? Does parsing have such high time complexity?

**Comment by rnveach** [comment]
I am good to approve with no concerns but this has to go at the same time as https://github.com/checkstyle/checkstyle/pull/15204 . I am not sure if it is safe to merge as release still hasn't happened and @romani was suppose to be looking into the issue already.

**Comment by mahfouz72** [comment]
Will we have 10.18.0 this month? I thought it would be postponed for next month due to the release issues

> I am not sure if it is safe to merge as release still hasn't happened

why not? why does it have to be after release?

**Comment by nrmancuso** [comment]
> Does parsing have such high time complexity?

Yes, this is our MOST performance sensitive area.

**Comment by nrmancuso** [comment]
>I am good to approve with no concerns but this has to go at the same time as https://github.com/checkstyle/checkstyle/pull/15204 .

To be clear, #15204 should go first.

**Comment by rnveach** [comment]
> Will we have 10.18.0 this month? I thought it would be postponed for next month due to the release issues

The last message we got from @romani was he would come back to the release issue in a week (already passed) and he would continue the release. He hasn't said it has been moved to the end of the month.

> why not? why does it have to be after release?

We want this and other PR to go together or as @nrmancuso says for the other to go first. I don't want to merge this and then find out a surprise release happened.

**Comment by romani** [comment]
Too much activity in projects, I deprioritized release in favor of unblocking activity in project. 
Should I change priority?

**Comment by romani** [comment]
It is safe to merge.
Don't forget to downgrade value in performance threshold.

**Comment by nrmancuso** [comment]
https://github.com/checkstyle/checkstyle/issues/15241

**Pull request merged**